File: genie/ambrosia/ambrosia.js
// Created:20110127
// By Jeff Connelly

// Ambrosia: an experimental push-style templating engine for JavaScript

// Inspirations: Amrita http://amrita.sourceforge.jp/
// other systems like HTML::Steamstress http://www.perlmonks.org/?node_id=674225
// and for theory http://www.cs.usfca.edu/~parrt/papers/mvc.templates.pdf

function ambrosia(node, value) {
    if (value instanceof Array) {
        // Clone node for each element of array
        for (var i = 0; i < value.length; i += 1) {
            var new_node = node.cloneNode(true); 
            // TODO: make id unique. or should we use classes for everything instead?

            node.parentNode.insertBefore(new_node, node)

            ambrosia(new_node, value[i]);
        }

        // Remove the template node
        node.parentNode.removeChild(node);
    } else if (value === null) {
        // Remove node
        node.parentNode.removeChild(node);
    } else if (value === undefined) {
        // No operation

    // https://developer.mozilla.org/en/JavaScript/Reference/Operators/Special_Operators/typeof_Operator
    // TODO: boolean, xml
    } else if (typeof value === "string" || typeof value === "number") {
        // Scalar text value
        node.textContent = value;
    } else if (typeof value === "function") {
        ambrosia(node, value.call());
    } else if (value instanceof AmbrosiaAttrList) {
        for (attr in value.attributes) { 
            if (value.attributes.hasOwnProperty(attr)) {
                node.setAttribute(attr, value.attributes[attr]);
            }
        }

        ambrosia(node, value.content);
    } else if (typeof value === "object") {
        // Nested
        for (var id in value) {
            if (value.hasOwnProperty(id)) {
                var next_node = node.querySelector("#" + id);
                if (!next_node) {
                    throw "no such node id: " + id + ", from " + next_node;
                }

                ambrosia(next_node, value[id]);
            }
        }
    } else {
        throw "ambrosia(" + node + ", " + value + "): unsupported data type: " + typeof value;
    }
}

function A(attributes, content) {
    if (typeof attributes === "string" || typeof attributes === "number") {
        // Also accept reversed order of parameters, sometimes more convenient
        return new AmbrosiaAttrList(content, attributes);
    }

    return new AmbrosiaAttrList(attributes, content);
}

function AmbrosiaAttrList(attributes, content) {
    this.attributes = attributes;
    this.content = content;
    return this;
}

File: genie/ambrosia/ambrosia.rb
#!/usr/bin/ruby
# Created:20110129
# By Jeff Connelly

# Ambrosia: an experimental push-style templating engine

require 'rubygems'
require 'nokogiri'


def ambrosia(html, data)
    root = Nokogiri::HTML::DocumentFragment.parse(html)

    root.ambrosia(data)

    return root.to_html
end

def expand_node(node, value)
    case value
    when Hash
        value.each do |key, next_value|
            next_node = node.at_css("\##{key}")
            if next_node.nil? 
                throw "expand_node(#{node}, #{value}): no such id: #{key}"
            end
            next_node.ambrosia(next_value)
        end
    when String
        node.content = value
    when Array
        value.each do |item|
            new_node = node.clone       # TODO: uniquify id
            node.parent.add_child(new_node)

            new_node.ambrosia(item)
        end
        node.remove
    when NilClass, FalseClass
        node.remove
    when TrueClass
        # no operation
    when AmbrosiaAttrList
        value.attributes.each do |attribute_name, attribute_value|
            node[attribute_name.to_s] = attribute_value 
        end
        node.ambrosia(value.content)
    else
        # Not a recognized type, so try by what it responds to
        if value.respond_to? :to_node
            value.to_node(node)
        elsif value.respond_to? :to_s
            node.content = value.to_s
        else
            throw "expand_node(#{node}, #{value}): unsupported data type: #{value.class}"
        end
    end
end

def A(attributes, content=true)
    if attributes.class == String || attributes.class == Fixnum
        # reversed order
        return AmbrosiaAttrList.new(content, attributes)
    end

    return AmbrosiaAttrList.new(attributes, content)
end

class AmbrosiaAttrList
    attr_accessor :attributes, :content
    def initialize(attributes, content)
        @attributes = attributes
        @content = content
    end
end


# Convenience methods to make it more OO
class Nokogiri::HTML::DocumentFragment
    def ambrosia(value)
        expand_node(self, value)
    end
end

class Nokogiri::XML::Element
    def ambrosia(value)
        expand_node(self, value)
    end
end


